### PR TITLE
dateutil first version

### PR DESCRIPTION
**Pull request by jukebox**

### CODE REVIEW

File: third_party/3/dateutil/parser.pyi [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+    JUMP = ...  # type: List[str]
+    WEEKDAYS = ...  # type: List[Tuple[str, str]]
+    MONTHS = ...  # type: List[Tuple[str, str]]
+    HMS = ...  # type: List[Tuple[str, str, str]]
+    AMPM = ...  # type: List[Tuple[str, str, str]]
+    UTCZONE = ...  # type: List[str]
+    PERTAIN = ...  # type: List[str]
+    TZOFFSET = ...  # type: Dict[str, int]
+
+    def __init__(self, dayfirst: bool=..., yearfirst: bool=...) -> None:
```
Comment (gvanrossum) [inline third_party/3/dateutil/parser.pyi]: Style suggestion: in stubs we prefer compactness and it's fine to put the "..." of the method body on the same line and omit the blank lines between methods.

File: third_party/3/dateutil/parser.pyi [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+        ...
+
+DEFAULTPARSER = ...  # type: parser
+
+
+def parse(timestr, parserinfo: parserinfo=..., **kwargs) -> datetime:
+    ...
+
+
+class _tzparser(object):
```
Comment (gvanrossum) [inline third_party/3/dateutil/parser.pyi]: Does this class not have any behavior?

**Comment by gvanrossum** [comment]
Hey, sorry for the slow response. I just realized I've never used dateutil so I don't really have any idea whether this stub is adequate or not.

Also, because it's third party, could you ask the dateutil author whether they are okay with having a dateutil stub in typeshed? See the note in README.md under "third_party".

**Comment by pganssle** [comment]
The original author of dateutil does not currently maintain the library, but I'm one of the principal maintainers and I'm fine with a stub being included.

File: third_party/3/dateutil/parser.pyi (hunk 2 above)
Comment (pganssle) [inline third_party/3/dateutil/parser.pyi]: `_tzparser` has behavior, but as far as I know it's exclusively used in [`tzstr`](https://github.com/dateutil/dateutil/blob/master/dateutil/tz/tz.py#L702) for parsing a string to a `_tzparser._result` object.

File: third_party/3/dateutil/parser.pyi [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+    def convertyear(self, year: int) -> int: ...
+    def validate(self, year: datetime) -> bool: ...
+
+
+class parser(object):
+    def __init__(self, info: parserinfo=...) -> None: ...
+
+    def parse(
+        self,
+        timestr: str,
```
Comment (pganssle) [inline third_party/3/dateutil/parser.pyi]: I'm not overly familiar with how type annotations work, but you may want to be aware that `timestr` duck-types the input. In Python 3.x will take `bytes`, `str` or stream objects that implement `read()`. The preferred input is probably `str`, though.

Comment (gvanrossum) [inline third_party/3/dateutil/parser.pyi]: Maybe a union of bytes, str and IO would work.

Comment (jukebox) [inline third_party/3/dateutil/parser.pyi]: done

**Comment by gvanrossum** [comment]
@jukebox: I don't understand what happened, but it looks like a whole bunch of merged commits ended up in your PR. Could you clean that up (maybe try a rebase) or start a new PR? The comment from https://github.com/python/typeshed/pull/133#issuecomment-208456565 indicates we can go ahead with this PR, but it's currently impossible to review due to the merge noise.

**Comment by jukebox** [comment]
is it ok now ?

**Comment by gvanrossum** [comment]
Yup, thanks! Does dateutil support Python 2 also?

**Comment by pganssle** [comment]
Dateutil supports python 2.6+ and 3.2+.

**Comment by pganssle** [comment]
@jukebox Are you going to continue filling out the rest of the stub? If not, I am probably in a pretty good position to flesh out most of the library. I'll just need a bit of time to familiarize myself with the exact notation and whatnot.

It seems that these stubs also cover the private interface, which makes sense, but makes things a bit more complicated, I would think, especially given #153.

**Comment by jukebox** [comment]
@pganssle i think i dont have time to continue the stub and i dont use other methods and old python versions.

**Comment by pganssle** [comment]
OK, I'll make a separate PR covering 2.x and expanding coverage to the whole library.

**Pull request merged**